File: code/ex924.cpp
class Solution {
public:
    int minMalwareSpread(vector<vector<int>>& graph, vector<int>& initial) {
        unordered_map<int, int> root;
        for (int i = 0; i < graph.size(); i++)    root[i] = i;
        
        // Union
        for (int i = 0; i < graph.size(); i++) {
            for (int j = 0; j < graph[0].size(); j++) {
                if (graph[i][j] == 1) {
                    uni(root, i, j);
                }
            }
        }
        
        // Get groups
        unordered_map<int, unordered_set<int>> group;
        for (int i = 0; i < graph.size(); i++) {
            group[find(root, i)].insert(i);
        }
        
        // Get the max group size which can be saved and get the initial to be deleted
        int MAX = -1, del = -1;
        for (auto entry : group) {
            unordered_set<int>& s = entry.second;
            int cnt = 0, d = -1;
            for (int init : initial) {
                if (s.find(init) != s.end()) {
                    cnt++;
                    d = init;
                }
            }
            
            int sz = s.size();
            if (cnt == 1) {    // Current group can be saved
                if (sz - MAX > 0) {
                    MAX = sz;
                    del = d;
                    
                } else if (s.size() - MAX == 0) {
                    del = min(del, d);
                }
            }
        }
        
        if (MAX == -1) {    // No group can be saved by removing one initial
            sort(initial.begin(), initial.end());
            return initial[0];
        } else {
            return del;
        }
    }
    
private:
    int find(unordered_map<int, int>& root, int p) {
        if (root[p] != p)
            root[p] = find(root, root[p]);
        return root[p];
    }
    
    void uni(unordered_map<int, int>& root, int a, int b) {
        int r1 = find(root, a), r2 = find(root, b);
        if (r1 <= r2)
            root[r2] = r1;
        else
            root[r1] = r2;
    }
};

File: code/ex924.java
class Solution {
    public int minMalwareSpread(int[][] graph, int[] initial) {
        int n = graph.length;
        int[] parent = new int[n];
        for (int i = 0; i < n; i++)
            parent[i] = i;
        
        // Union
        for (int i = 0; i < n; i++) {
            for (int j = i + 1; j < n; j++) {
                if (graph[i][j] == 1)
                    uni(parent, i, j);
            }
        }
        
        // Get group size
        Map<Integer, Integer> group = new HashMap<>();
        for (int i = 0; i < n; i++) {
            int root = find(parent, i);
            group.put(root, group.getOrDefault(root, 0) + 1);
        }
        
        int max = 0, res = -1;
        Arrays.sort(initial);
        for (int init : initial) {
            int members = group.get(find(parent ,init));
            if (members > max) {
                max = members;
                res = init;
            }
        }
        
        return res;
    }
    
    private void uni(int[] parent, int a, int b) {
        int ra = find(parent, a), rb = find(parent, b);
        if (ra <= rb)    parent[rb] = ra;
        else    parent[ra] = rb;
    }
    
    private int find(int[] parent, int a) {
        if (parent[a] != a)
            parent[a] = find(parent, parent[a]);
        return parent[a];
    }
}